### PR TITLE
[KnownBits] Make `{s,u}{add,sub}_sat` optimal

### PR DESCRIPTION
**Pull request by goldsteinn**
Changes are:
    1) Make signed-overflow detection optimal
    2) For signed-overflow, try to rule out direction even if we can't
       totally rule out overflow.
    3) Intersect add/sub assuming no overflow with possible overflow
       clamping values as opposed to add/sub without the assumption.

### CODE REVIEW

**Comment by goldsteinn** [comment]
ping

**Comment by RKSimon** [comment]
please can you rebase to get the CI to run again?

**Comment by goldsteinn** [comment]
Rebased

File: llvm/lib/Support/KnownBits.cpp [hunk 1]
```diff
@@ -610,28 +610,78 @@ static KnownBits computeForSatAddSub(bool Add, bool Signed,
...
+      Overflow = false;
+    else {
+      // Check if we may overflow. If we can't rule out overflow then check if
+      // we can rule out a direction at least.
+      KnownBits UnsignedLHS = LHS;
+      KnownBits UnsignedRHS = RHS;
+      UnsignedLHS.One.clearSignBit();
+      UnsignedLHS.Zero.setSignBit();
+      UnsignedRHS.One.clearSignBit();
+      UnsignedRHS.Zero.setSignBit();
```
Comment (RKSimon) [inline llvm/lib/Support/KnownBits.cpp:638]: How are we allowed to do this? Please can you add a short description in the comments?

**Review by RKSimon — COMMENTED** [review]
@jayfoad any thoughts?

**Comment by jayfoad** [comment]
> jayfoad any thoughts?

Only that if the exhaustive unit test passes then the logic must be OK, so no objections from me. I don't have time to try to understand the patch deeply now.

**Review by RKSimon — APPROVED** [review]

**Pull request merged**